File: leetcode/dp/300. Longest Increasing Subsequence/js/main.js
// DP Solution
// Time: O(N^2)
// Space: O(N)
const lengthOfLIS = nums => {
  if (nums.length == 0) return 0;
  let n = nums.length;

  let dp = Array(n).fill(1);
  let maxLen = 1;

  for (let i = 0; i < n; i++) {
    for (let j = i + 1; j < n; j++) {
      if (nums[i] < nums[j]) {
        dp[j] = Math.max(dp[j], dp[i] + 1);
      }

      if (dp[j] > maxLen) {
        maxLen = dp[j];
      }
    }
  }

  return maxLen;
};

// Time O(N^2)
// Space O(N^2)
const lengthOfLIS_II = nums => {
  if (nums.length == 0) return 0;
  let n = nums.length;

  let dp = Array(n + 1)
    .fill(0)
    .map(() => Array(n).fill(null));

  return dfs(-1, 0);

  function dfs(prevIndex, currIndex) {
    if (currIndex == n) {
      return 0;
    }

    if (dp[prevIndex + 1][currIndex] != null) {
      return dp[prevIndex + 1][currIndex];
    }

    let taken = 0;
    if (prevIndex < 0 || nums[currIndex] > nums[prevIndex]) {
      taken = 1 + dfs(currIndex, currIndex + 1);
    }

    let nottaken = dfs(prevIndex, currIndex + 1);

    dp[prevIndex + 1][currIndex] = Math.max(taken, nottaken);

    return dp[prevIndex + 1][currIndex];
  }
};

File: leetcode/dp/300. Longest Increasing Subsequence/py/main.py
# Time O(NLogN)
# Space O(N)
class Solution:
    def solve(self, nums):
        if not nums:
            return 0
        
        def search(nums, target):
            lo, hi = 0, len(nums) - 1
            
            while lo < hi:
                mid = (lo + hi) >> 1
                
                if target > nums[mid]:
                    lo = mid + 1
                else:
                    hi = mid
            
            return lo
            
        n = len(nums)
        stack = []
        
        for num in nums:
            if not stack or num > stack[-1]:
                stack.append(num)
            else:
                pos = search(stack, num)
                
                stack[pos] = num
                
        return len(stack)
        

# Time O(N^2)
# Space O(N)
class Solution_II:
    def lengthOfLIS(self, nums):
        if not len(nums):
          return 0
        
        n = len(nums)
        dp = [1] * n
        
        res = 1
        
        for i in range(0, n):
            for j in range(i + 1, n):
                if (nums[i] < nums[j]):
                    dp[j] = max(dp[j], dp[i] + 1)
                    
                res = max(res, dp[j])
                
        return res
 

# Time O(N^2)
# Space O(N^2)
class Solution_III:
    def lengthOfLIS(self, nums):
        if not len(nums):
          return 0
        
        n = len(nums)
        dp = [[None] * n for _ in range(n)]

        def dfs(prev_index, next_index):
          if next_index >= n:
            return 0
          
          if dp[prev_index + 1][next_index] is not None:
            return dp[prev_index + 1][next_index]

          res = 0

          if prev_index < 0 or nums[prev_index] < nums[next_index]:
            res = max(res, 1 + dfs(next_index, next_index + 1))
            
          res = max(res,  dfs(prev_index, next_index + 1))
            
          dp[prev_index + 1][next_index] = res

          return dp[prev_index + 1][next_index]
              
        return dfs(-1, 0)
